sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+namespace Azure.Storage.Common.DataMovement
+{
+    /// <summary>
+    /// FilesystemScanner class
+    /// </summary>
+    public static class FilesystemScanner
+    {
+        /// <summary>
+        /// Enumerates all files pointed to by the provided path, including those in subdirectories (if path is a directory).
+        /// </summary>
+        /// <param name="path">Filesystem location.</param>
+        /// <returns>Enumerable list of absolute paths containing all relevant files the user has permission to access.</returns>
+        public static IEnumerable<string> ScanLocation(string path)
+        {
+            // Path type is ambiguous at start
+            bool isDirectory = false;
+
+            try
+            {
+                // Make sure we're dealing with absolute path
+                path = Path.GetFullPath(path);
+
+                // Check if path points to a directory
+                if ((File.GetAttributes(path) & FileAttributes.Directory) == FileAttributes.Directory)
+                {
+                    isDirectory = true;
+                }
+            }
+            catch
+            {
+                // If there's an error here, there aren't any valid entries to scan at the
+                // given path; either the path is invalid/nonexistant, or it isn't accessible by the user. In this case,
+                // don't return anything.
+                yield break;
+            }
+
+            // If we're given a directory, parse its children recursively
+            if (isDirectory)
+            {
+                // Create a queue of folders to enumerate files from, starting with provided path
+                Queue<string> folders = new Queue<string>();
+                folders.Enqueue(path);
+
+                while (folders.Count > 0)
+                {
+                    // Grab a folder from the queue
+                    string dir = folders.Dequeue();
+
+                    // Try to enumerate and queue all subdirectories of the current folder
+                    try
+                    {
+                        foreach (string subdir in Directory.EnumerateDirectories(dir))
+                        {
+                            folders.Enqueue(subdir);
+                        }
+                    }
+                    // If we lack permissions to enumerate, skip this folder
+                    catch
+                    {
+                        continue;
```

Comment:
If we lack the permissions we should somehow log this or maybe throw an exception. Maybe let's add a TODO here to add a logging.

Reply:
TODO added for logging; I set it to throw if we get an exception trying to enumerate the parent folder, but skip and continue for subfolders.

I'm thinking that if a user wants to enumerate a folder, we'll first either be able to read the given path or not, and if we can, we'll either be able to read all or some of the subfolders inside (provided they exist). A user might want to only enumerate if every file is possible to enumerate, or just enumerate whatever is possible at the path supplied. I set the scanner to work for the latter case for now, but maybe we can make this an option to be set by the user (single argument or in options object)?

(Should I set it to the former case for now instead, or is this fine?)

Reply:
Yeah that's a good idea, we will just follow what the user wants to do in the case. This is fine.